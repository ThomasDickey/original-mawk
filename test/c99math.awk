#
# tests for C99 equivalent Arithmetic (if compiled into mawk)
#
# These tests will FAIL if -DSTDC_MATHERR is defined
#
BEGIN {
  Arr["PI"] = 3.1415926535897932384626433832795
  Arr["NZero"] = -0.0
  Arr["Zero"] = 0.0
  Arr["IntVal"] = 42
  Arr["DblVal"] = 42.42
  Arr["NaN"] = log(-3)
  Arr["Inf"] = tgamma(12345678)
  Arr["NInf"] = log2(0)

  for (x in Arr) {
    print "\n###############",Arr[x],"#####################"
    print x, "abs            = ", abs(Arr[x])
    print x, "sin            = ", sin(Arr[x])
    print x, "cos            = ", cos(Arr[x])
    print x, "tan            = ", tan(Arr[x])
    print x, "asin           = ", asin(Arr[x])
    print x, "acos           = ", acos(Arr[x])
    print x, "atan           = ", atan(Arr[x])
    print x, "sinh           = ", sinh(Arr[x])
    print x, "cosh           = ", cosh(Arr[x])
    print x, "tanh           = ", tanh(Arr[x])
    print x, "asinh          = ", asinh(Arr[x])
    print x, "acosh          = ", acosh(Arr[x])
    print x, "atanh          = ", atanh(Arr[x])
    print x, "hypot(x,0.1)   = ", hypot(Arr[x],0.1)
    print x, "log            = ", log(Arr[x])
    print x, "log10          = ", log10(Arr[x])
    print x, "log2           = ", log2(Arr[x])
    print x, "log1p          = ", log1p(Arr[x])
    print x, "logb           = ", logb(Arr[x])
    print x, "ilogb          = ", ilogb(Arr[x])
    print x, "exp            = ", exp(Arr[x])
    print x, "exp2           = ", exp2(Arr[x])
    print x, "expm1          = ", expm1(Arr[x])
    print x, "ldexp(x,2)     = ", ldexp(Arr[x],2)
    print x, "pow(x,10)      = ", pow(Arr[x],10)
    print x, "cbrt           = ", cbrt(Arr[x])

    print x, "mod(x,2)       = ", mod(Arr[x],2)
    print x, "int            = ", int(Arr[x])
    print x, "ceil           = ", ceil(Arr[x])
    print x, "floor          = ", floor(Arr[x])
    print x, "trunc          = ", trunc(Arr[x])
    print x, "round          = ", round(Arr[x])
    print x, "min(x,2)       = ", min(Arr[x],2)
    print x, "max(x,2)       = ", max(Arr[x],2)
    print x, "erf            = ", erf(Arr[x])
    print x, "erfc           = ", erfc(Arr[x])
    print x, "lgamma         = ", lgamma(Arr[x])
    print x, "tgamma         = ", tgamma(Arr[x])
    print x, "bessel_j0      = ", bessel_j0(Arr[x])
    print x, "bessel_j1      = ", bessel_j1(Arr[x])
    print x, "bessel_jn(3,x) = ", bessel_jn(3,Arr[x])
    print x, "bessel_y0      = ", bessel_y0(Arr[x])
    print x, "bessel_y1      = ", bessel_y1(Arr[x])
    print x, "bessel_yn(3,x) = ", bessel_yn(3,Arr[x])

    print x, "compl          = ", compl(Arr[x])
    print x, "lshift(x,2)    = ", lshift(Arr[x],2)
    print x, "rshift(x,2)    = ", rshift(Arr[x],2)
    print x, "xor(x,0xff)    = ", xor(Arr[x],0xff)
    print x, "or(x,0xff)     = ", or(Arr[x],0xff)
    print x, "and(x,0xff)    = ", and(Arr[x],0xff)
  }
  print "##########################################"
}
