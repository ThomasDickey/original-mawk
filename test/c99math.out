
############### 0 #####################
Zero abs            =  0
Zero sin            =  0
Zero cos            =  1
Zero tan            =  0
Zero asin           =  0
Zero acos           =  1.5708
Zero atan           =  0
Zero sinh           =  0
Zero cosh           =  1
Zero tanh           =  0
Zero asinh          =  0
Zero acosh          =  -nan
Zero atanh          =  0
Zero hypot(x,0.1)   =  0.1
Zero log            =  -inf
Zero log10          =  -inf
Zero log2           =  -inf
Zero log1p          =  0
Zero logb           =  -inf
Zero ilogb          =  -2.14748e+09
Zero exp            =  1
Zero exp2           =  1
Zero expm1          =  0
Zero ldexp(x,2)     =  0
Zero pow(x,10)      =  0
Zero cbrt           =  0
Zero mod(x,2)       =  0
Zero int            =  0
Zero ceil           =  0
Zero floor          =  0
Zero trunc          =  0
Zero round          =  0
Zero min(x,2)       =  0
Zero max(x,2)       =  2
Zero erf            =  0
Zero erfc           =  1
Zero lgamma         =  inf
Zero tgamma         =  inf
Zero bessel_j0      =  1
Zero bessel_j1      =  0
Zero bessel_jn(3,x) =  0
Zero bessel_y0      =  -inf
Zero bessel_y1      =  -inf
Zero bessel_yn(3,x) =  -inf
Zero compl          =  4.29497e+09
Zero lshift(x,2)    =  0
Zero rshift(x,2)    =  0
Zero xor(x,0xff)    =  255
Zero or(x,0xff)     =  255
Zero and(x,0xff)    =  0

############### inf #####################
Inf abs            =  inf
Inf sin            =  -nan
Inf cos            =  -nan
Inf tan            =  -nan
Inf asin           =  nan
Inf acos           =  nan
Inf atan           =  1.5708
Inf sinh           =  inf
Inf cosh           =  inf
Inf tanh           =  1
Inf asinh          =  inf
Inf acosh          =  inf
Inf atanh          =  -nan
Inf hypot(x,0.1)   =  inf
Inf log            =  inf
Inf log10          =  inf
Inf log2           =  inf
Inf log1p          =  inf
Inf logb           =  inf
Inf ilogb          =  2147483647
Inf exp            =  inf
Inf exp2           =  inf
Inf expm1          =  inf
Inf ldexp(x,2)     =  inf
Inf pow(x,10)      =  inf
Inf cbrt           =  inf
Inf mod(x,2)       =  -nan
Inf int            =  inf
Inf ceil           =  inf
Inf floor          =  inf
Inf trunc          =  inf
Inf round          =  inf
Inf min(x,2)       =  2
Inf max(x,2)       =  inf
Inf erf            =  1
Inf erfc           =  0
Inf lgamma         =  inf
Inf tgamma         =  inf
Inf bessel_j0      =  0
Inf bessel_j1      =  0
Inf bessel_jn(3,x) =  0
Inf bessel_y0      =  0
Inf bessel_y1      =  0
Inf bessel_yn(3,x) =  0
Inf compl          =  0
Inf lshift(x,2)    =  4.29497e+09
Inf rshift(x,2)    =  1073741823
Inf xor(x,0xff)    =  4.29497e+09
Inf or(x,0xff)     =  4.29497e+09
Inf and(x,0xff)    =  255

############### 42 #####################
IntVal abs            =  42
IntVal sin            =  -0.916522
IntVal cos            =  -0.399985
IntVal tan            =  2.29139
IntVal asin           =  nan
IntVal acos           =  nan
IntVal atan           =  1.54699
IntVal sinh           =  8.69637e+17
IntVal cosh           =  8.69637e+17
IntVal tanh           =  1
IntVal asinh          =  4.43096
IntVal acosh          =  4.43068
IntVal atanh          =  -nan
IntVal hypot(x,0.1)   =  42.0001
IntVal log            =  3.73767
IntVal log10          =  1.62325
IntVal log2           =  5.39232
IntVal log1p          =  3.7612
IntVal logb           =  5
IntVal ilogb          =  5
IntVal exp            =  1.73927e+18
IntVal exp2           =  4.39805e+12
IntVal expm1          =  1.73927e+18
IntVal ldexp(x,2)     =  168
IntVal pow(x,10)      =  1.70802e+16
IntVal cbrt           =  3.47603
IntVal mod(x,2)       =  0
IntVal int            =  42
IntVal ceil           =  42
IntVal floor          =  42
IntVal trunc          =  42
IntVal round          =  42
IntVal min(x,2)       =  2
IntVal max(x,2)       =  42
IntVal erf            =  1
IntVal erfc           =  0
IntVal lgamma         =  114.034
IntVal tgamma         =  3.34525e+49
IntVal bessel_j0      =  -0.114739
IntVal bessel_j1      =  -0.0459939
IntVal bessel_jn(3,x) =  0.0567129
IntVal bessel_y0      =  -0.044625
IntVal bessel_y1      =  0.114216
IntVal bessel_yn(3,x) =  -0.109448
IntVal compl          =  4.29497e+09
IntVal lshift(x,2)    =  168
IntVal rshift(x,2)    =  10
IntVal xor(x,0xff)    =  213
IntVal or(x,0xff)     =  255
IntVal and(x,0xff)    =  42

############### -inf #####################
NInf abs            =  inf
NInf sin            =  -nan
NInf cos            =  -nan
NInf tan            =  -nan
NInf asin           =  nan
NInf acos           =  nan
NInf atan           =  -1.5708
NInf sinh           =  -inf
NInf cosh           =  inf
NInf tanh           =  -1
NInf asinh          =  -inf
NInf acosh          =  -nan
NInf atanh          =  -nan
NInf hypot(x,0.1)   =  inf
NInf log            =  nan
NInf log10          =  nan
NInf log2           =  nan
NInf log1p          =  -nan
NInf logb           =  inf
NInf ilogb          =  2147483647
NInf exp            =  0
NInf exp2           =  0
NInf expm1          =  -1
NInf ldexp(x,2)     =  -inf
NInf pow(x,10)      =  inf
NInf cbrt           =  -inf
NInf mod(x,2)       =  -nan
NInf int            =  -inf
NInf ceil           =  -inf
NInf floor          =  -inf
NInf trunc          =  -inf
NInf round          =  -inf
NInf min(x,2)       =  -inf
NInf max(x,2)       =  2
NInf erf            =  -1
NInf erfc           =  2
NInf lgamma         =  inf
NInf tgamma         =  nan
NInf bessel_j0      =  0
NInf bessel_j1      =  0
NInf bessel_jn(3,x) =  0
NInf bessel_y0      =  nan
NInf bessel_y1      =  nan
NInf bessel_yn(3,x) =  nan
NInf compl          =  4.29497e+09
NInf lshift(x,2)    =  0
NInf rshift(x,2)    =  0
NInf xor(x,0xff)    =  255
NInf or(x,0xff)     =  255
NInf and(x,0xff)    =  0

############### 0 #####################
NZero abs            =  0
NZero sin            =  0
NZero cos            =  1
NZero tan            =  0
NZero asin           =  0
NZero acos           =  1.5708
NZero atan           =  0
NZero sinh           =  0
NZero cosh           =  1
NZero tanh           =  0
NZero asinh          =  0
NZero acosh          =  -nan
NZero atanh          =  0
NZero hypot(x,0.1)   =  0.1
NZero log            =  -inf
NZero log10          =  -inf
NZero log2           =  -inf
NZero log1p          =  0
NZero logb           =  -inf
NZero ilogb          =  -2.14748e+09
NZero exp            =  1
NZero exp2           =  1
NZero expm1          =  0
NZero ldexp(x,2)     =  0
NZero pow(x,10)      =  0
NZero cbrt           =  0
NZero mod(x,2)       =  0
NZero int            =  0
NZero ceil           =  0
NZero floor          =  0
NZero trunc          =  0
NZero round          =  0
NZero min(x,2)       =  0
NZero max(x,2)       =  2
NZero erf            =  0
NZero erfc           =  1
NZero lgamma         =  inf
NZero tgamma         =  -inf
NZero bessel_j0      =  1
NZero bessel_j1      =  0
NZero bessel_jn(3,x) =  0
NZero bessel_y0      =  -inf
NZero bessel_y1      =  -inf
NZero bessel_yn(3,x) =  -inf
NZero compl          =  4.29497e+09
NZero lshift(x,2)    =  0
NZero rshift(x,2)    =  0
NZero xor(x,0xff)    =  255
NZero or(x,0xff)     =  255
NZero and(x,0xff)    =  0

############### 3.14159 #####################
PI abs            =  3.14159
PI sin            =  1.22465e-16
PI cos            =  -1
PI tan            =  -1.22465e-16
PI asin           =  nan
PI acos           =  nan
PI atan           =  1.26263
PI sinh           =  11.5487
PI cosh           =  11.592
PI tanh           =  0.996272
PI asinh          =  1.8623
PI acosh          =  1.81153
PI atanh          =  -nan
PI hypot(x,0.1)   =  3.14318
PI log            =  1.14473
PI log10          =  0.49715
PI log2           =  1.6515
PI log1p          =  1.42108
PI logb           =  1
PI ilogb          =  1
PI exp            =  23.1407
PI exp2           =  8.82498
PI expm1          =  22.1407
PI ldexp(x,2)     =  12.5664
PI pow(x,10)      =  93648
PI cbrt           =  1.46459
PI mod(x,2)       =  1.14159
PI int            =  3
PI ceil           =  4
PI floor          =  3
PI trunc          =  3
PI round          =  3
PI min(x,2)       =  2
PI max(x,2)       =  3.14159
PI erf            =  0.999991
PI erfc           =  8.87615e-06
PI lgamma         =  0.827695
PI tgamma         =  2.28804
PI bessel_j0      =  -0.304242
PI bessel_j1      =  0.284615
PI bessel_jn(3,x) =  0.333458
PI bessel_y0      =  0.328366
PI bessel_y1      =  0.358873
PI bessel_yn(3,x) =  -0.48607
PI compl          =  4.29497e+09
PI lshift(x,2)    =  12
PI rshift(x,2)    =  0
PI xor(x,0xff)    =  252
PI or(x,0xff)     =  255
PI and(x,0xff)    =  3

############### 42.42 #####################
DblVal abs            =  42.42
DblVal sin            =  -0.999964
DblVal cos            =  0.00849907
DblVal tan            =  -117.656
DblVal asin           =  nan
DblVal acos           =  nan
DblVal atan           =  1.54723
DblVal sinh           =  1.32355e+18
DblVal cosh           =  1.32355e+18
DblVal tanh           =  1
DblVal asinh          =  4.44091
DblVal acosh          =  4.44063
DblVal atanh          =  -nan
DblVal hypot(x,0.1)   =  42.4201
DblVal log            =  3.74762
DblVal log10          =  1.62757
DblVal log2           =  5.40667
DblVal log1p          =  3.77092
DblVal logb           =  5
DblVal ilogb          =  5
DblVal exp            =  2.64711e+18
DblVal exp2           =  5.88427e+12
DblVal expm1          =  2.64711e+18
DblVal ldexp(x,2)     =  169.68
DblVal pow(x,10)      =  1.88672e+16
DblVal cbrt           =  3.48757
DblVal mod(x,2)       =  0.42
DblVal int            =  42
DblVal ceil           =  43
DblVal floor          =  42
DblVal trunc          =  42
DblVal round          =  42
DblVal min(x,2)       =  2
DblVal max(x,2)       =  42.42
DblVal erf            =  1
DblVal erfc           =  0
DblVal lgamma         =  115.601
DblVal tgamma         =  1.603e+50
DblVal bessel_j0      =  -0.0861389
DblVal bessel_j1      =  -0.0881222
DblVal bessel_jn(3,x) =  0.0958529
DblVal bessel_y0      =  -0.087101
DblVal bessel_y1      =  0.0851184
DblVal bessel_yn(3,x) =  -0.0765268
DblVal compl          =  4.29497e+09
DblVal lshift(x,2)    =  168
DblVal rshift(x,2)    =  10
DblVal xor(x,0xff)    =  213
DblVal or(x,0xff)     =  255
DblVal and(x,0xff)    =  42

############### nan #####################
NaN abs            =  nan
NaN sin            =  nan
NaN cos            =  nan
NaN tan            =  nan
NaN asin           =  nan
NaN acos           =  nan
NaN atan           =  nan
NaN sinh           =  nan
NaN cosh           =  nan
NaN tanh           =  nan
NaN asinh          =  nan
NaN acosh          =  nan
NaN atanh          =  nan
NaN hypot(x,0.1)   =  nan
NaN log            =  nan
NaN log10          =  nan
NaN log2           =  nan
NaN log1p          =  nan
NaN logb           =  nan
NaN ilogb          =  -2.14748e+09
NaN exp            =  nan
NaN exp2           =  nan
NaN expm1          =  nan
NaN ldexp(x,2)     =  nan
NaN pow(x,10)      =  nan
NaN cbrt           =  nan
NaN mod(x,2)       =  nan
NaN int            =  nan
NaN ceil           =  nan
NaN floor          =  nan
NaN trunc          =  nan
NaN round          =  nan
NaN min(x,2)       =  2
NaN max(x,2)       =  2
NaN erf            =  nan
NaN erfc           =  nan
NaN lgamma         =  nan
NaN tgamma         =  nan
NaN bessel_j0      =  nan
NaN bessel_j1      =  nan
NaN bessel_jn(3,x) =  nan
NaN bessel_y0      =  nan
NaN bessel_y1      =  nan
NaN bessel_yn(3,x) =  nan
NaN compl          =  4.29497e+09
NaN lshift(x,2)    =  0
NaN rshift(x,2)    =  0
NaN xor(x,0xff)    =  255
NaN or(x,0xff)     =  255
NaN and(x,0xff)    =  0
##########################################
